numpy/_core/src/multiarray/multiarraymodule.c

```diff
@@ -3274,161 +3275,163 @@ PyArray_Where(PyObject *condition, PyObject *x, PyObject *y)
     NPY_cast_info cast_info = {.func = NULL};
 
     ax = (PyArrayObject*)PyArray_FROM_O(x);
+    if (ax == NULL) {
+        goto fail;
+    }
     ay = (PyArrayObject*)PyArray_FROM_O(y);
-    if (ax == NULL || ay == NULL) {
+    if (ay == NULL) {
         goto fail;
     }
-    else {
-        npy_uint32 flags = NPY_ITER_EXTERNAL_LOOP | NPY_ITER_BUFFERED |
-                           NPY_ITER_REFS_OK | NPY_ITER_ZEROSIZE_OK;
-        PyArrayObject * op_in[4] = {
-            NULL, arr, ax, ay
-        };
-        npy_uint32 op_flags[4] = {
-            NPY_ITER_WRITEONLY | NPY_ITER_ALLOCATE | NPY_ITER_NO_SUBTYPE,
-            NPY_ITER_READONLY,
-            NPY_ITER_READONLY | NPY_ITER_ALIGNED,
-            NPY_ITER_READONLY | NPY_ITER_ALIGNED
-        };
-        PyArray_Descr * common_dt = PyArray_ResultType(2, &op_in[0] + 2,
-                                                       0, NULL);
-        PyArray_Descr * op_dt[4] = {common_dt, PyArray_DescrFromType(NPY_BOOL),
-                                    common_dt, common_dt};
-        NpyIter * iter;
-        NPY_BEGIN_THREADS_DEF;
-
-        if (common_dt == NULL || op_dt[1] == NULL) {
-            Py_XDECREF(op_dt[1]);
-            Py_XDECREF(common_dt);
-            goto fail;
-        }
-        iter =  NpyIter_MultiNew(4, op_in, flags,
-                                 NPY_KEEPORDER, NPY_UNSAFE_CASTING,
-                                 op_flags, op_dt);
-        Py_DECREF(op_dt[1]);
-        Py_DECREF(common_dt);
-        if (iter == NULL) {
-            goto fail;
-        }
+    npy_mark_tmp_array_if_pyscalar(x, ax, NULL);
+    npy_mark_tmp_array_if_pyscalar(y, ay, NULL);
+
+    npy_uint32 flags = NPY_ITER_EXTERNAL_LOOP | NPY_ITER_BUFFERED |
+                        NPY_ITER_REFS_OK | NPY_ITER_ZEROSIZE_OK;
+    PyArrayObject * op_in[4] = {
+        NULL, arr, ax, ay
+    };
+    npy_uint32 op_flags[4] = {
+        NPY_ITER_WRITEONLY | NPY_ITER_ALLOCATE | NPY_ITER_NO_SUBTYPE,
+        NPY_ITER_READONLY,
+        NPY_ITER_READONLY | NPY_ITER_ALIGNED,
+        NPY_ITER_READONLY | NPY_ITER_ALIGNED
+    };
+    common_dt = PyArray_ResultType(2, &op_in[2], 0, NULL);
+    if (common_dt == NULL) {
+        goto fail;
+    }
+    PyArray_Descr * op_dt[4] = {common_dt, PyArray_DescrFromType(NPY_BOOL),
```

Comment:
Couldn't hurt to add a `\* creating a built in dtype like this cannot fail *\` above this line.

Reply:
OK, we really should probably just have a name for it...